feat(core): add enter and leave animation instructions

thePunderWoman opened the pull request:
This adds the instructions to support enter and leave animations on nodes.




## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman committed:
feat(core): add enter and leave animation instructions

This adds the instructions to support enter and leave animations on nodes.

JeanMeche commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,114 @@
+
+import {Injectable} from './di';
+
+export type AnimationCallbackEvent = {target: Element; animationComplete: Function};
+export type AnimationFunction = (event: AnimationCallbackEvent) => void;
```
Both lack `@public-api 20.2` and some quick doc. (can be extended later if we add a dedicated guide page)

thePunderWoman replied:
Thanks for the heads up. I've added them.

michael-small commented on packages/core/test/acceptance/animation_spec.ts:
```diff
@@ -0,0 +1,628 @@
+        selector: 'test-cmp',
+        styles: multiple,
+        template:
+          '<div>@if (show()) {<p [animate.leave]="classArray" #el>I should slide in</p>}</div>',
```
```suggestion
          '<div>@if (show()) {<p [animate.leave]="classArray" #el>I should slide out</p>}</div>',
```

thePunderWoman commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,129 @@
+  animate(el: Element, removeFn: Function): void {
+    if (!this.outElements.has(el)) return;
+    const details = this.outElements.get(el)!;
+    const timeout = setTimeout(() => removeFn(), 4000);
```
Also to quickly answer your question here, if you're using an animation function (rather than classes) for `animate.leave`, you are required to call the complete function to tell the framework when to remove the element. The feedback on the RFC was to have a timeout for safety, and so I went with the same timeout duration as cross document view transitions. So after 4 seconds, the element gets removed.

AndrewKushnir commented on goldens/public-api/core/index.api.md:
```diff
@@ -667,6 +679,23 @@ export class ElementRef<T = any> {
 }
 
+// @public
+export class ElementRegistry {
```
Do we need this symbol in the public API, was it intentionally exposed?

thePunderWoman replied:
It's no longer in the public API.

AndrewKushnir commented on packages/core/test/bundling/forms_reactive/bundle.golden_symbols.json:
```diff
@@ -96,6 +97,7 @@
       "EffectRefImpl",
       "EffectScheduler",
       "ElementRef",
+      "ElementRegistry",
```
This looks unexpected, should we try making it tree-shakable (maybe via using lightweight injection token in the non-tree-shakable part and providing an actual implementation once the feature is brought it via special instructions)?

thePunderWoman replied:
I'm not sure how we would be able to do this since there's no provider function for Animations.

AndrewKushnir replied:
+1, left a comment here: https://github.com/angular/angular/pull/62682#discussion_r2216806087

thePunderWoman replied:
I mentioned this in another comment, but the features subsystem is per component / per directive. Whereas the animations element registry only has to be enabled once. So if we use this subsystem, we end up adding bytes to every component / directive rather than adding bytes with the ElementRegistry. So I'm personally not sure this approach is the right one to trim bytes.

thePunderWoman replied:
Regardless, I was able to add a feature to make this tree shakeable.

AndrewKushnir commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,137 @@
+ * It's intended to prevent people from accidentally forgetting to call the removal
+ * function in their callback.
+ */
+const ANIMATION_TIMEOUT = 4000;
```
nit: more descriptive const name:
```suggestion
const MAX_ANIMATION_DURATION = 4000;
```

AndrewKushnir commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,137 @@
+    if (!details.classFns) {
+      details.classFns = [resolver];
+    } else {
+      details.classFns?.push(resolver);
```
nit: do we need `?` here? It looks like the type should be inferred correctly and should not require it:
```suggestion
      details.classFns.push(resolver);
```

AndrewKushnir commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,137 @@
+    if (!this.outElements.has(el)) return;
+    const details = this.outElements.get(el)!;
+    const timeout = setTimeout(() => removeFn(), ANIMATION_TIMEOUT);
+    details.animateFn(() => removeFn(timeout));
```
Should we also call `clearTimeout` here?

thePunderWoman replied:
The timeout is cleared inside the `animateFn`. We wouldn't want to clear it here as that would prevent the behavior of the max timeout we're looking for.

AndrewKushnir commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,409 @@
+  const nativeElement = getNativeByTNode(tNode, lView) as HTMLElement;
+
+  if ((nativeElement as Node).nodeType !== Node.ELEMENT_NODE) {
+    return ɵɵanimateEnter; // SSR, do nothing and allow chaining
```
It looks like the comment was copied from the previous code branch.

AndrewKushnir commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,409 @@
+  const tNode = getCurrentTNode()!;
+  const nativeElement = getNativeByTNode(tNode, lView) as HTMLElement;
+
+  if ((nativeElement as Node).nodeType !== Node.ELEMENT_NODE) {
```
Quick question: do we detect cases when animation is used on non-DOM elements, e.g. `<ng-container>`, `<ng-template>`, `<ng-content>` in compiler? Or this is a protection against host bindings (where we do not know whether the target element is a DOM node)?

thePunderWoman replied:
I don't know that we check in the compiler about non-DOM elements. So this is protection in case someone throws `animate.*` on something that's not supported.

AndrewKushnir commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,409 @@
+  const classes = typeof value === 'function' ? value() : value;
+  let activeEnterClasses: string[] | null = classes instanceof Array ? classes : null;
+  if (typeof classes === 'string') {
+    activeEnterClasses = classes
+      .trim()
+      .split(WS_REGEXP)
+      .filter((k) => k);
```
nit: we might already have a util fn to split classes like this. We probably use similar logic for `[class]` bindings. cc @crisbeto

thePunderWoman replied:
I actually pulled this code directly out of `ng_class.ts`. If we have a util function, I'm happy to switch to it.

AndrewKushnir commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,409 @@
+  } else {
+    if (ngDevMode) {
+      throw new RuntimeError(
+        901, // Placeholder error code for animateEnter
```
Should we move this code to the corresponding `RuntimeError` enum?

thePunderWoman commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,409 @@
+  }
+
+  if (animationsDisabled) {
+    // add microtask
```
I was thinking about this exact comment and how terrible it is last night. Yes, this needs more info.

thePunderWoman commented:
/gemini review

AndrewKushnir commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,409 @@
+): LongestAnimation | undefined {
+  if (!(event.target instanceof Element)) return;
+  const nativeElement = event.target;
+  if (typeof nativeElement.getAnimations === 'function') {
```
Got it, thanks. My thinking was to have this check somewhere at the module level, e.g. check if `document.body.getAnimations` is present and just use this info everywhere. The problem that in this case well be bypassing `renderer` completely, which may not be great (but likely in practice it'd not be a big problem). Maybe there are some other ways to do it? WDYT?

Another idea is that we can move this check from this utility function into an instruction, so that we just exit early (similar to SSR checks).

AndrewKushnir commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,151 @@
+ */
+@Injectable({providedIn: 'root'})
+export class ElementRegistry {
+  private outElements = new Map<Element, AnimationDetails>();
```
Should it be a `WeakMap` instead?

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -317,6 +326,15 @@ class DefaultDomRenderer2 implements Renderer2 {
   }
 
   removeChild(_parent: any, oldChild: any): void {
+    if (this.registry && this.registry.has(oldChild as Element)) {
+      this.registry.animate(oldChild, (timeout: ReturnType<typeof setTimeout>) => {
+        clearTimeout(timeout);
+        this.registry?.remove(oldChild);
+        oldChild.remove();
+      });
+      return;
+    }
+    // child was removed
     oldChild.remove();
```
We should try to minimize the amount of non-tree-shakable code here (in favor of having more code inside of the registry itself):
```suggestion
    if (!this.registry?.animateRemoval(oldChild, () => oldChild.remove())) {
      // Either animation registry is not present (i.e. a component doesn't use animations)
      // or that element was not in the registry. If that's the case - remove the element here.
      // Otherwise, it'd be removed by the animation logic at the end of the animation.
      oldChild.remove();
    }
```

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -149,6 +150,8 @@ export class DomRendererFactory2 implements RendererFactory2, OnDestroy {
     @Inject(TracingService)
     @Optional()
     private readonly tracingService: TracingService<TracingSnapshot> | null = null,
+    @Inject(ElementRegistry)
+    private readonly registry: ElementRegistry | null,
```
Based on the current implementation, the `ElementRegistry` would always be present (since it's provided in `root`).

I like the idea that @JeanMeche proposed: we can use "features" subsystem to bring in `ElementRegistry` in a tree-shakable way. Another benefit would be that the `ElementRegistry` would not need to be exposed via public API, since renderers would not need to have it as an argument. We'll be able to retrieve it from the component type (see `type: RendererType2` in the `createRenderer` method below). The "feature" would basically be a function that given an injector, returns an instance of the `ElementRegistry` (e.g. calls `injector.get(ElementRegistry, ...)`).

thePunderWoman replied:
The downside to the features subsystem is that it's on a per component level. We don't need to enable the element registry per component. We just need to do it once. So we end up with feature code on every component / directive to enable the element registry once. It feels like there should be another way we can do this, but I'll try it.

thePunderWoman replied:
This has now successfully been removed from the public API.

thePunderWoman commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,151 @@
+    let timeoutId: ReturnType<typeof setTimeout>;
+    let called = false;
+    const remove = () => {
+      if (called) return;
+      called = true;
```
This was specifically suggested by Gemini to prevent race conditions. You can view the comment Gemini left to see why.

crisbeto commented on packages/core/src/animation.ts:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  /** Used when animate.leave is only applying classes */
+  addClasses(details: AnimationDetails, classes: string | string[]): void {
```
nit: maybe something like `trackClasses` would be more descriptive here? `addClasses` made me think that it's actually adding the classes to the DOM.

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+  const tNode = getCurrentTNode()!;
+  const nativeElement = getNativeByTNode(tNode, lView) as HTMLElement;
+
+  if ((nativeElement as Node).nodeType !== Node.ELEMENT_NODE) {
```
I remember at some point that `Node` wasn't available during SSR. You should be able to get the type of the DOM node through `tNode.type`.

thePunderWoman replied:
We don't actually run this code during SSR. There's an early escape, but I can switch it.

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+
+  // We only need to add these event listeners if there are actual classes to apply
+  if (activeClasses && activeClasses.length > 0) {
+    renderer.listen(nativeElement, 'animationstart', handleAnimationStart, {once: true});
```
Should we track these for cleaning up as well? Otherwise we may end up leaking if the element is removed before the animation has started. Also I'm guessing that either `animationstart` or `transitionstart` will fire, but likely not both.

crisbeto replied:
Should we bind these outside of the `NgZone`?

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+  if (activeClasses && activeClasses.length > 0) {
+    renderer.listen(nativeElement, 'animationstart', handleAnimationStart, {once: true});
+    renderer.listen(nativeElement, 'transitionstart', handleAnimationStart, {once: true});
+    cleanupFns.push(renderer.listen(nativeElement, 'animationend', handleInAnimationEnd));
```
Would it make sense to bind these from insert the `animationstart`/`transitionstart` event? That way we can bind only the relevant end event.

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+    // The animations will only be disabled in a test environment, and adding a microtask here
+    // will allow the tests to be able to tick forward to resolve the next phase of animation
+    // in their tests.
+    new Promise<void>((resolve) => {
```
I think the way we schedule these in other places is `Promise.resolve().then(...)` or is this supposed to run earlier?

thePunderWoman replied:
Ah, I can switch to that.

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+  const tNode = getCurrentTNode()!;
+  const nativeElement = getNativeByTNode(tNode, lView) as HTMLElement;
+
+  if ((nativeElement as Node).nodeType !== Node.ELEMENT_NODE) {
```
Same note here about checking the node type.

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+    return ɵɵanimateEnterListener;
+  }
+
+  value.call(lView[CONTEXT], {target: nativeElement, animationComplete: () => {}});
```
Should we have a shared no-op function so we don't need to recreate the `animationComplete` each time the instruction is called?

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+function cancelAnimationsIfRunning(element: HTMLElement): void {
+  if (areAnimationSupported) {
+    for (const animation of element.getAnimations()) {
+      if (animation.playState === 'running') {
```
Aren't we running the risk here that we can cancel one of the user's animations that we didn't initiate?

crisbeto commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,468 @@
+): LongestAnimation | undefined {
+  if (!(event.target instanceof Element)) return;
+  const nativeElement = event.target;
+  if (areAnimationSupported) {
```
nit: maybe combine this check with the `event.target instanceof Element` above? It would help reduce the nesting.

crisbeto commented on packages/core/test/acceptance/animation_spec.ts:
```diff
@@ -0,0 +1,677 @@
+
+(function () {
+  // these tests are only meant to be run within the DOM (for now)
+  if (isNode) return;
```
nit: you can reduce the nesting in the entire file with something like this:

```
describe('Animation', () => {
  if (isNode) {
    it('should pass', () => expect(true).toBe(true));
  }
  
  // Other tests...
});
```

crisbeto commented on packages/core/test/acceptance/animation_spec.ts:
```diff
@@ -0,0 +1,677 @@
+          show = signal(true);
+        }
+
+        TestBed.configureTestingModule({
```
You don't need this call if the test component is standalone.

thePunderWoman committed:
refactor(compiler): Tree shake Element Registry

This creates a feature to detect usages of animate.leave and only enables the element removal registry when necessary

AndrewKushnir commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,475 @@
+  };
+
+  ngZone.runOutsideAngular(() => {
+    renderer.listen(el, 'animationstart', handleAnimationStart, {once: true});
+    renderer.listen(el, 'transitionstart', handleAnimationStart, {once: true});
+    renderer.listen(el, 'animationend', handleOutAnimationEnd);
+    renderer.listen(el, 'transitionend', handleOutAnimationEnd);
```
Should we also skip adding listeners if animations are disabled?

AndrewKushnir reviewed (commented):
@thePunderWoman thanks for making the `ElementRegistry` tree-shakable 👍 

I've left a few minor comments and I also wanted to propose that we check whether we can use `areAnimationSupported` to exit a bit earlier (e.g. in the main instructions vs inside of util functions later).

thePunderWoman commented on goldens/public-api/core/index.api.md:
```diff
@@ -81,6 +81,18 @@ export interface AfterViewInit {
+export type AnimationFunction = (event: AnimationCallbackEvent) => void;
+
+// @public
+export const ANIMATIONS_DISABLED: InjectionToken<boolean>;
```
IMHO, this feels like too much to me. Maybe we consider it if we see more options that would need configuring down the road. For now I think an injection token for test environments is enough as there's nothing needed to provide in non-test environments.

thePunderWoman commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,475 @@
+export function ɵɵanimateEnter(value: string | Function): typeof ɵɵanimateEnter {
+  performanceMarkFeature('NgAnimateEnter');
+
+  if (typeof ngServerMode !== 'undefined' && ngServerMode) {
```
Yeah that makes sense. I've added it to all the instructions here now.

thePunderWoman commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,475 @@
+  const injector = lView[INJECTOR]!;
+
+  // Assume ElementRegistry and ANIMATIONS_DISABLED are injectable services.
+  const elementRegistry = getRegistry()!;
```
Yeah, that sounds like a good idea. I've added it.

atscott commented on packages/core/src/render3/instructions/animation.ts:
```diff
@@ -6,18 +6,476 @@
+    // The animations will only be disabled in a test environment, and adding a microtask here
+    // will allow the tests to be able to tick forward to resolve the next phase of animation
+    // in their tests.
+    Promise.resolve().then(() => {
```
I think this should run outside the Angular zone too, right?

atscott replied:
So to be fair, this will cause tests to behave differently than production (thanks NgZone...) by causing an extra change detection. For better or for worse (definitely for worse), we have observed tests trying to count CD cycles. There could be other cases that this affects as well. For example, the listener callbacks, which execute outside the Angular zone, may make changes to application state. If this isn't done with `ngZone.run`, `markForCheck`, etc. production code will _not_ know to trigger change detection but the test code will.

AndrewKushnir reviewed (approved):
LGTM, thanks @thePunderWoman 👍 

Just left some minor comments, but none of them are blocking. Let's discuss the testing story as a followup.

crisbeto reviewed (approved):
Reviewed-for: public-api

JeanMeche reviewed (approved):
reviewed-for: public-api

kirjs commented:
This PR was merged into the repository by commit 882522c1ad95a451945b648f5feb14136f11b3c2.

The changes were merged into the following branches: main
